tulip/spec/prioritized_safety.py

```diff
@@ -0,0 +1,167 @@
+#
+# 1. Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#
+# 3. Neither the name of the copyright holder(s) nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL CALTECH
+# OR THE CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+# USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+# OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+# SUCH DAMAGE.
+
+from itertools import product
+from tulip.transys.automata import FiniteStateAutomaton as FA
+
+
+class FAWithPriority(object):
+    """A class for defining a rule, represented by a nondeterministic finite automaton,
+    with priority"""
+
+    def __init__(self, fa, priority, level):
+        assert isinstance(fa, FA)
+        assert isinstance(priority, int)
+        assert isinstance(level, int)
+
+        self._fa = fa
+        self._priority = priority
+        self._level = level
+
+    def priority(self):
+        """Get the priority of this rule"""
+        return self._priority
+
+    def automaton(self):
+        """Get the automaton of this rule"""
+        return self._fa
+
+    def level(self):
+        """Get the level of this rule"""
+        return self._level
+
+
+class PrioritizedSpecification(object):
+    """A class for defining a prioritized safety specification"""
+
+    def __init__(self):
+        self._Psi = []
+        self.atomic_propositions = []
+
+    def __getitem__(self, key):
+        assert key >= 0
+        level = 0
+        while level < len(self._Psi) and key >= len(self._Psi[level]):
+            key -= len(self._Psi[level])
+            level += 1
+        if level < len(self._Psi) and key < len(self._Psi[level]):
+            return self._Psi[level][key]
+        raise IndexError("index out of range")
+
+    def __iter__(self):
+        self._iter_level = 0
+        self._iter_index = 0
+        return self
+
+    def __next__(self):
+        while self._iter_level < len(self._Psi) and self._iter_index >= len(
+            self._Psi[self._iter_level]
+        ):
```

Comment:
Is this PEP8 compliant?  Suggest running `pycodestyle`

Reply:
It was generated from `black` and validated with flake8.

Reply:
As far as I know, it is PEP8 compliant, but some readers might find it difficult to read compared with the style of folding the line at the `and`, e.g.,
```python
while (self._iter_level < len(self._Psi)
       and self._iter_index >= len(self._Psi[self._iter_level])):
```